### ISSUE TITLE
Tips to speed up training

### ISSUE THREAD

**Issue by shivakanthsujit** [NVlabs/vla0#21]
I'd been profiling the training part of the code and noticed that the are some easy ways to speed up the code (sometimes by upto **2.5x**)

In the forward pass of the model, it calculates the sysuser_len per example in a loop. This loop is the slowest part of the forward pass. And it's because the processor is called on each example individually instead of in a batch. By doing it in a batch, you get nearly a 60% speedup in training. This could be done as below

```py
            sysuser_convs = [e[:-1] for e in examples]
            sysuser_texts = [
                self.processor.apply_chat_template(
                    conv,
                    tokenize=False,
                    add_vision_id=self.add_vision_id,
                )
                for conv in sysuser_convs
            ]
            sysuser_imgs = [process_vision_info(conv)[0] for conv in sysuser_convs]
            sysuser_inputs = self.processor(
                text=sysuser_texts,
                images=sysuser_imgs,
                return_tensors="pt",
                padding=True,
            )
            # number of non-padding tokens per sample
            sysuser_lens = sysuser_inputs["attention_mask"].sum(dim=1)
            sysuser_lens = sysuser_lens + 3
```
And then you can use sysuser_lens in the loop to get the masking indices for the labels.

Further more, it is possible to get the sysuser_lens without even calling the processor. This is because the sysuser_len is the total tokens - action_tokens - assistant_end_message_tokens (=2 in Qwen's case). Computing the action tokens is much less expensive compared to the whole sequence. You can get them the following way.

```py
            batch_seq_len = labels.size(1)
            batch_nonpad_tokens = model_inputs["attention_mask"].sum(dim=1)  # (bs,)
            action_token_lens = [
                len(self.processor.tokenizer(action_txt[i], add_special_tokens=False)["input_ids"])
                for i in range(bs)
            ]
            action_token_lens = np.array(action_token_lens)
            sysuser_lens = batch_nonpad_tokens.cpu() - action_token_lens - 2
```
This means the tokeniser is called only once on a batch (for getting `model_inputs`), allowing for about a 40% speedup. Together with the flash attention fix, you can speedup training by **3x** compared to the original code. 

If you feel it is useful, I can submit a PR incorporating these changes.

**Comment by shivakanthsujit** [comment]
Oh damn, it's 1.35s/it with the pinned torch/transformers versions. And the slow operations I'd identified before are now much faster (0.7s vs 0.05s) so the change I suggested isn't needed anymore actually. The fix I mentioned above produces a negligible speedup now. The batch processing make the code more compact, but it doesn't affect training speed.  

Thanks for the fix!

**Comment by imankgoyal** [comment]
Sounds good. Closing this issue for now then.
